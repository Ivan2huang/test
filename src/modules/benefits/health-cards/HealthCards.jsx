import React from 'react';
import PropTypes from 'prop-types';
import HealthCard from './HealthCard';
import { EMPLOYEE, TIER_3, TERMINATED } from '../constant';
import { IsEmployee } from '../../../helpers/roles';

const HealthCards = ({ memberProfile, loaded }) => {
  if (Object.keys(memberProfile.policy).length === 0) {
    return '';
  }

  const {
    memberId,
    relationships,
    membershipNumber: mainMembershipNumber,
    planId: mainPlanId,
    healthCards,
    coPayments,
    certificateNumber,
    policy,
    role,
    category,
    firstName,
    lastName,
  } = memberProfile;
  const mainFullName = `${lastName} ${firstName}`;
  const { policyNumber } = policy;
  const {
    code: insurerCode,
    eHealthCardName: insurerName,
    cardType: insurerType,
  } = policy.insurer;

  const memberHealthCard = healthCards.find(healthCard => {
    return healthCard.memberId === memberId;
  });

  const filteredHealthCards = [];
  if (IsEmployee(role)) {
    healthCards.forEach(h => {
      if (h.memberId !== memberId) {
        const dependantInfo = relationships.find(
          r => r.memberId === h.memberId && r.status !== TERMINATED,
        );
        if (dependantInfo) {
          dependantInfo.cardType = h.type;
          filteredHealthCards.push(dependantInfo);
        }
      }
    });
  }

  const getCoPayment = planId => {
    return category === TIER_3 ? null : coPayments[planId];
  };

  return (
    loaded && (
      <>
        {memberHealthCard && (
          <HealthCard
            fullName={mainFullName}
            membershipNumber={mainMembershipNumber}
            coPayment={
              role === EMPLOYEE
                ? coPayments[mainPlanId]
                : getCoPayment(mainPlanId)
            }
            cardType={memberHealthCard.type}
            certificateNumber={certificateNumber}
            insurerCode={insurerCode}
            insurerName={insurerName}
            insurerType={insurerType}
            policyNumber={policyNumber}
          />
        )}
        {filteredHealthCards.map((healthCard, index) => {
          const {
            firstName: memberFirstName,
            lastName: memberLastName,
            membershipNumber,
            planId,
            cardType,
          } = healthCard;
          return (
            <HealthCard
              /* eslint-disable-next-line react/no-array-index-key */
              key={`healthCard-${index}`}
              fullName={`${memberLastName} ${memberFirstName}`}
              membershipNumber={membershipNumber}
              coPayment={getCoPayment(planId)}
              cardType={cardType}
              certificateNumber={certificateNumber}
              insurerCode={insurerCode}
              insurerName={insurerName}
              insurerType={insurerType}
              policyNumber={policyNumber}
            />
          );
        })}
      </>
    )
  );
};

HealthCards.propTypes = {
  memberProfile: PropTypes.shape({
    firstName: PropTypes.string.isRequired,
    lastName: PropTypes.string.isRequired,
    memberId: PropTypes.string.isRequired,
    membershipNumber: PropTypes.string.isRequired,
    planId: PropTypes.number,
    relationships: PropTypes.arrayOf(
      PropTypes.shape({
        firstName: PropTypes.string.isRequired,
        lastName: PropTypes.string.isRequired,
        memberId: PropTypes.string.isRequired,
        planId: PropTypes.number,
      }),
    ).isRequired,
    coPayments: PropTypes.shape({}).isRequired,
    healthCards: PropTypes.arrayOf(
      PropTypes.shape({
        memberId: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired,
      }).isRequired,
    ).isRequired,
    category: PropTypes.string.isRequired,
  }).isRequired,
  loaded: PropTypes.bool.isRequired,
};

export default HealthCards;
