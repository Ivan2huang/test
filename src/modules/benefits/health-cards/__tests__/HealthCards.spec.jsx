/* eslint-disable no-undef */
import React from 'react';
import { render } from '@testing-library/react';

import HealthCards from '../HealthCards';
import withIntl from '../../../../i18n/withIntlProvider';

jest.mock(
  '../HealthCard',
  // eslint-disable-next-line react/prop-types
  () => props => (
    <div>
      HealthCard Component
      <span>{mockPropsCapture(props)}</span>
    </div>
  ),
);

jest.mock(
  '../../../../uiComponents/Typography',
  // eslint-disable-next-line react/prop-types
  () => ({ children, props }) => (
    <div {...props}>
      Typography Component
      {children}
    </div>
  ),
);

jest.mock('../../../../helpers/helpers', () => ({
  formatMessage: jest.fn((intl, id, defaultMessage) => defaultMessage),
}));

describe('HealthCards Component', () => {
  const props = {
    memberProfile: {
      firstName: 'William',
      lastName: 'Brown',
      memberId: '3',
      membershipNumber: '123',
      certificateNumber: 'CERT123',
      role: 'Employee',
      planId: 1,
      category: 'Tier 1',
      relationships: [
        {
          firstName: 'William',
          lastName: 'Black',
          memberId: '111',
          membershipNumber: '123',
          planId: 1,
        },
        {
          firstName: 'William',
          lastName: 'White',
          memberId: '112',
          membershipNumber: '124',
          planId: 1,
          status: 'Terminated',
        },
      ],
      coPayments: { 1: { GP: 10, SP: 10, PHY: 10 } },
      healthCards: [
        { memberId: '3', type: 'PRIMARY' },
        { memberId: '111', type: 'SECONDARY' },
        { memberId: '112', type: 'SECONDARY' },
      ],
      policy: {
        policyNumber: 'PN123',
        expiryDate: '12-3-2020',
        insurer: {
          code: 'IC123',
          name: 'ABC insurer',
          eHealthCardName: 'test eHealth card name',
          cardType: 'HSBC',
        },
      },
    },
    getMemberProfile: jest.fn(),
    setIsLoaded: jest.fn(),
    loaded: true,
  };

  it('should match the snapshot', () => {
    const Component = withIntl(HealthCards);
    const { container } = render(<Component {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot when member role not employee', () => {
    const Component = withIntl(HealthCards);
    const newProps = {
      ...props,
      memberProfile: {
        ...props.memberProfile,
        role: 'Dependent',
      },
    };
    const { container } = render(<Component {...newProps} />);

    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot when not render copayment section', () => {
    const Component = withIntl(HealthCards);
    const newProps = {
      ...props,
      memberProfile: {
        ...props.memberProfile,
        category: 'Tier 3',
      },
    };
    const { container } = render(<Component {...newProps} />);

    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot when not rendering health cards', () => {
    props.memberProfile.healthCards = [{ memberId: '111', type: 'PRIMARY' }];
    const Component = withIntl(HealthCards);
    const { container } = render(<Component {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot when not provide policy data', () => {
    const nextProps = {
      memberProfile: {
        ...props.memberProfile,
        policy: [],
      },
      loaded: true,
    };
    const Component = withIntl(HealthCards);
    const { container } = render(<Component {...nextProps} />);

    expect(container).toMatchSnapshot();
  });
});
