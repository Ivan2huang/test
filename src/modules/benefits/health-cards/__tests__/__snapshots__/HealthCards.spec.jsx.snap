// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HealthCards Component should match the snapshot 1`] = `
<div>
  <div>
    HealthCard Component
    <span>
      fullName = Brown William
      membershipNumber = 123
      coPayment = [object Object]
      cardType = PRIMARY
      certificateNumber = CERT123
      insurerCode = IC123
      insurerName = test eHealth card name
      insurerType = HSBC
      policyNumber = PN123
    </span>
  </div>
  <div>
    HealthCard Component
    <span>
      fullName = Black William
      membershipNumber = 123
      coPayment = [object Object]
      cardType = SECONDARY
      certificateNumber = CERT123
      insurerCode = IC123
      insurerName = test eHealth card name
      insurerType = HSBC
      policyNumber = PN123
    </span>
  </div>
</div>
`;

exports[`HealthCards Component should match the snapshot when member role not employee 1`] = `
<div>
  <div>
    HealthCard Component
    <span>
      fullName = Brown William
      membershipNumber = 123
      coPayment = [object Object]
      cardType = PRIMARY
      certificateNumber = CERT123
      insurerCode = IC123
      insurerName = test eHealth card name
      insurerType = HSBC
      policyNumber = PN123
    </span>
  </div>
</div>
`;

exports[`HealthCards Component should match the snapshot when not provide policy data 1`] = `
<div>
  
</div>
`;

exports[`HealthCards Component should match the snapshot when not render copayment section 1`] = `
<div>
  <div>
    HealthCard Component
    <span>
      fullName = Brown William
      membershipNumber = 123
      coPayment = [object Object]
      cardType = PRIMARY
      certificateNumber = CERT123
      insurerCode = IC123
      insurerName = test eHealth card name
      insurerType = HSBC
      policyNumber = PN123
    </span>
  </div>
  <div>
    HealthCard Component
    <span>
      fullName = Black William
      membershipNumber = 123
      coPayment = null
      cardType = SECONDARY
      certificateNumber = CERT123
      insurerCode = IC123
      insurerName = test eHealth card name
      insurerType = HSBC
      policyNumber = PN123
    </span>
  </div>
</div>
`;

exports[`HealthCards Component should match the snapshot when not rendering health cards 1`] = `
<div>
  <div>
    HealthCard Component
    <span>
      fullName = Black William
      membershipNumber = 123
      coPayment = [object Object]
      cardType = PRIMARY
      certificateNumber = CERT123
      insurerCode = IC123
      insurerName = test eHealth card name
      insurerType = HSBC
      policyNumber = PN123
    </span>
  </div>
</div>
`;
